dev-tools/xtask/src/a4x2_deploy.rs

```diff
@@ -0,0 +1,472 @@
+    sh.copy_file(
+        "a4x2-ssh-key.pub",
+        env.a4x2_dir.join("cargo-bay/g0/root_authorized_keys"),
+    )?;
+    sh.copy_file(
+        "a4x2-ssh-key.pub",
+        env.a4x2_dir.join("cargo-bay/g1/root_authorized_keys"),
+    )?;
+    sh.copy_file(
+        "a4x2-ssh-key.pub",
+        env.a4x2_dir.join("cargo-bay/g2/root_authorized_keys"),
+    )?;
+    sh.copy_file(
+        "a4x2-ssh-key.pub",
+        env.a4x2_dir.join("cargo-bay/g3/root_authorized_keys"),
+    )?;
+
+    Ok(())
+}
+
+fn try_launch_a4x2(sh: &Shell, env: &Environment) -> Result<()> {
+    let _popdir = sh.push_dir(&env.a4x2_dir);
+
+    cmd!(sh, "pfexec ./a4x2 launch").run()?;
+
+    // XXX Is there a better way to do this?
+    let ce_addr_json =
+        cmd!(sh, "./a4x2 exec ce 'ip -4 -j addr show enp0s10'").read()?;
+
+    // Translated from this jq query:
+    // .[0].addr_info[] | select(.dynamic == true) | .local
+    let ce_addr_json: Value = serde_json::from_str(&ce_addr_json)?;
+    let customer_edge_addr = &ce_addr_json[0]["addr_info"]
+        .as_array()
+        .unwrap()
+        .iter()
+        .find(|v| v["dynamic"] == Value::Bool(true))
+        .ok_or(anyhow!("failed to find customer edge addr"))?["local"]
+        .as_str()
+        .unwrap();
+
+    // XXX Im told that pinging the gateway from inside the sleds is no longer
+    // necessary so I'm leaving it out for now. We'll see if that's true.
+    // So far it seems to be.
+
+    cmd!(sh, "pfexec route add 198.51.100.0/24 {customer_edge_addr}").run()?;
+
+    // Not sure how this IP is fixed, but it is
+    let api_url = "http://198.51.100.23";
+    println!("polling control plane for signs of life for up to 25 minutes");
+
+    // Print the date for the logs' benefit
+    let _ = cmd!(sh, "date").run();
+
+    // Timeout = (retries / 2) minutes
+    // XXX this is an arbitrary timeout. Should it be configurable? Skippable?
+    let mut retries = 50;
+
+    // The important thing here is to do an HTTP request with timeout to the
+    // control plane API endpoint. If the server is up, we'll get the page you'd
+    // expect for an unauthenticated user. If not, then the request will either
+    // - fail
+    // - stall
+    //
+    // Stalling is why we do a timeout, and we need to handle timeout both at
+    // the TCP level and the HTTP level. `curl -m 5` will hard-out after 5
+    // seconds no matter what's going on.
+    //
+    // That 5 seconds is arbitrary, but it's been working well over in
+    // rackletteadm, from which this logic is copied.
+    //
+    // We could replace curl with rust code, but make sure we do the same thing!
+    while retries > 0 && cmd!(sh, "curl -s -m 5 {api_url}").run().is_err() {
+        retries -= 1;
+        thread::sleep(time::Duration::from_secs(25));
+
+        if retries % 5 == 0 {
+            // I see no reason to error while printing the date
+            let _ = cmd!(sh, "date").run();
+        }
+    }
```

Comment:
I wonder if you can use the same code that the end-to-end tests use.  I seem to recall them doing the same thing in Rust.

Reply:
If we do that, then it's worth factoring that out so we don't have to build e.g. our Diesel code to build this binary. (And if we do that, we should definitely put this into an external binary.)